{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.chdir(\"/home/ec2-user/scVI/\")\n",
    "os.getcwd()\n",
    "use_cuda = True\n",
    "\n",
    "from scvi.dataset.dataset import GeneExpressionDataset\n",
    "from scvi.dataset.pbmc import Dataset10X, PbmcDataset\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "from scvi.inference.posterior import *\n",
    "from scvi.metrics.clustering import select_indices_evenly, clustering_scores, entropy_batch_mixing\n",
    "from scipy import sparse\n",
    "from scvi.models.vae import VAE\n",
    "from scvi.models.scanvi import SCANVI\n",
    "\n",
    "from scvi.inference import UnsupervisedTrainer, SemiSupervisedTrainer\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from scvi.inference.posterior import get_bayes_factors\n",
    "from scvi.metrics.clustering import select_indices_evenly\n",
    "import torch\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def auc_score_threshold(gene_set, bayes_factor, gene_symbols):\n",
    "    # put ones on the genes from the gene_set\n",
    "    true_labels = np.array([g in gene_set for g in gene_symbols])\n",
    "    estimated_score = np.abs(bayes_factor)\n",
    "    indices = np.isfinite(estimated_score)\n",
    "    return roc_auc_score(true_labels[indices], estimated_score[indices])\n",
    "\n",
    "\n",
    "def WeightedAccuracy(y,y_pred,cell_types):\n",
    "    res = dict()\n",
    "    for i in np.unique(y):\n",
    "        res[cell_types[i]] = (np.mean(y_pred[y == i] == i), sum(y==i))\n",
    "    return(res)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File data/gene_info_pbmc.csv already downloaded\n",
      "File data/pbmc_metadata.pickle already downloaded\n",
      "File data/10X/pbmc8k/filtered_gene_bc_matrices.tar.gz already downloaded\n",
      "Preprocessing dataset\n",
      "Finished preprocessing dataset\n",
      "Downsampling from 33694 to 21425 genes\n",
      "Downsampling from 8381 to 8381 cells\n",
      "File data/10X/pbmc4k/filtered_gene_bc_matrices.tar.gz already downloaded\n",
      "Preprocessing dataset\n",
      "Finished preprocessing dataset\n",
      "Downsampling from 33694 to 19773 genes\n",
      "Downsampling from 4340 to 4340 cells\n",
      "Keeping 19039 genes\n",
      "Downsampling from 12721 to 11990 cells\n",
      "Downsampling from 19039 to 19036 genes\n",
      "Downsampling from 11990 to 11990 cells\n",
      "Downsampling from 19036 to 3346 genes\n",
      "Downsampling from 11990 to 11990 cells\n",
      "Downsampling from 11990 to 7982 cells\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/numpy/matrixlib/defmatrix.py:536: RuntimeWarning: Mean of empty slice.\n",
      "  return N.ndarray.mean(self, axis, dtype, out, keepdims=True)._collapse(axis)\n",
      "/home/ec2-user/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/numpy/core/_methods.py:73: RuntimeWarning: invalid value encountered in true_divide\n",
      "  ret, rcount, out=ret, casting='unsafe', subok=False)\n",
      "/home/ec2-user/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/numpy/matrixlib/defmatrix.py:604: RuntimeWarning: Degrees of freedom <= 0 for slice\n",
      "  return N.ndarray.var(self, axis, dtype, out, ddof, keepdims=True)._collapse(axis)\n",
      "/home/ec2-user/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/numpy/core/_methods.py:105: RuntimeWarning: invalid value encountered in true_divide\n",
      "  arrmean, rcount, out=arrmean, casting='unsafe', subok=False)\n",
      "/home/ec2-user/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/numpy/core/_methods.py:125: RuntimeWarning: invalid value encountered in true_divide\n",
      "  ret, rcount, out=ret, casting='unsafe', subok=False)\n"
     ]
    }
   ],
   "source": [
    "pbmc = PbmcDataset()\n",
    "de_data  = pbmc.de_metadata\n",
    "pbmc.update_cells(pbmc.batch_indices.ravel()==0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File data/10X/fresh_68k_pbmc_donor_a/filtered_gene_bc_matrices.tar.gz already downloaded\n",
      "Preprocessing dataset\n",
      "Finished preprocessing dataset\n"
     ]
    }
   ],
   "source": [
    "# pbmc.labels = pbmc.labels.reshape(len(pbmc),1)\n",
    "\n",
    "donor = Dataset10X('fresh_68k_pbmc_donor_a')\n",
    "donor.gene_names = donor.gene_symbols\n",
    "import os\n",
    "if not os.path.isfile('data/10X/fresh_68k_pbmc_donor_a/68k_pbmc_barcodes_annotation.tsv'):\n",
    "    import urllib.request\n",
    "    annotation_url = 'https://raw.githubusercontent.com/10XGenomics/single-cell-3prime-paper/master/pbmc68k_analysis/68k_pbmc_barcodes_annotation.tsv'\n",
    "    urllib.request.urlretrieve(annotation_url, 'data/10X/fresh_68k_pbmc_donor_a/68k_pbmc_barcodes_annotation.tsv')\n",
    "\n",
    "annotation = pd.read_csv('data/10X/fresh_68k_pbmc_donor_a/68k_pbmc_barcodes_annotation.tsv',sep='\\t')\n",
    "cellid1 = donor.barcodes\n",
    "temp = cellid1.join(annotation)\n",
    "assert all(temp[0]==temp['barcodes'])\n",
    "\n",
    "donor.cell_types,donor.labels = np.unique(temp['celltype'],return_inverse=True)\n",
    "donor.labels = donor.labels.reshape(len(donor.labels),1)\n",
    "donor.cell_types = np.array([ 'CD14+ Monocytes','B cells','CD34 cells', 'CD4 T cells 2','CD4 T cells Regulatory',\n",
    "                             'CD4 T cells Naive','CD4 T cells Memory','NK cells',\n",
    "                            'CD8 T cells',  'CD8 T cells Naive', 'Dendritic Cells'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Keeping 3343 genes\n"
     ]
    }
   ],
   "source": [
    "all_dataset = GeneExpressionDataset.concat_datasets(pbmc, donor)\n",
    "# Now resolve the Gene symbols to properly work with the DE\n",
    "all_gene_symbols = donor.gene_symbols[\n",
    "    np.array(\n",
    "        [np.where(donor.gene_names == x)[0][0] for x in list(all_dataset.gene_names)]\n",
    "    )]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(145, 145, 137, 138)\n",
      "['Other' 'CD4 T cells Regulatory' 'CD8 T cells Naive' 'B cells'\n",
      " 'CD8 T cells' 'Megakaryocytes' 'CD4 T cells Memory' 'CD4 T cells Naive'\n",
      " 'CD4 T cells' 'CD34 cells' 'FCGR3A+ Monocytes' 'NK cells' 'CD4 T cells 2'\n",
      " 'Dendritic Cells' 'CD14+ Monocytes']\n",
      "['Unnamed: 0' 'ENSG' 'GS' 'CD_logFC' 'CD_AveExpr' 'CD_t' 'CD_P.Value'\n",
      " 'CD_adj.P.Val' 'CD_B' 'BDC_logFC' 'BDC_AveExpr' 'BDC_t' 'BDC_P.Value'\n",
      " 'BDC_adj.P.Val' 'BDC_B' 'BDC2_logFC' 'BDC2_AveExpr' 'BDC2_t'\n",
      " 'BDC2_P.Value' 'BDC2_adj.P.Val' 'BDC2_B']\n"
     ]
    }
   ],
   "source": [
    "#####################################################################\n",
    "# Gene sets 1\n",
    "############################################################################\n",
    "path_geneset = \"Additional_Scripts/genesets.txt\"\n",
    "geneset_matrix = np.loadtxt(path_geneset, dtype=np.str)[:, 2:]\n",
    "CD4_TCELL_VS_BCELL_NAIVE, CD8_TCELL_VS_BCELL_NAIVE, CD8_VS_CD4_NAIVE_TCELL, NAIVE_CD8_TCELL_VS_NKCELL \\\n",
    "    = [set(geneset_matrix[i:i + 2, :].flatten()) & set(all_gene_symbols) for i in [0, 2, 4, 6]]\n",
    "\n",
    "# these are the length of the positive gene sets for the DE\n",
    "print((len(CD4_TCELL_VS_BCELL_NAIVE), len(CD8_TCELL_VS_BCELL_NAIVE),\n",
    "       len(CD8_VS_CD4_NAIVE_TCELL), len(NAIVE_CD8_TCELL_VS_NKCELL)))\n",
    "\n",
    "print(all_dataset.cell_types)\n",
    "\n",
    "comparisons = [\n",
    "    ['CD4 T cells', 'B cells'],\n",
    "    ['CD8 T cells', 'B cells'],\n",
    "    ['CD8 T cells', 'CD4 T cells'],\n",
    "    ['CD8 T cells', 'NK cells']\n",
    "               ]\n",
    "\n",
    "\n",
    "gene_sets = [CD4_TCELL_VS_BCELL_NAIVE,\n",
    "             CD8_TCELL_VS_BCELL_NAIVE,\n",
    "             CD8_VS_CD4_NAIVE_TCELL,\n",
    "             NAIVE_CD8_TCELL_VS_NKCELL]\n",
    "\n",
    "#####################################################################\n",
    "# Gene sets 2\n",
    "############################################################################\n",
    "print(de_data.columns.values)\n",
    "CD = de_data['CD_adj.P.Val']\n",
    "BDC = de_data['BDC_adj.P.Val']\n",
    "BDC2 = de_data['BDC2_adj.P.Val']\n",
    "CD = np.asarray(de_data['GS'][CD<0.05])\n",
    "BDC = np.asarray(de_data['GS'][BDC<0.05])\n",
    "BDC2 = np.asarray(de_data['GS'][BDC2<0.05])\n",
    "\n",
    "gene_sets = [set(CD) & set(all_gene_symbols),\n",
    "             set(BDC)& set(all_gene_symbols),\n",
    "             set(BDC2) &  set(all_gene_symbols)]\n",
    "\n",
    "comparisons = [\n",
    "    ['CD8 T cells', 'CD4 T cells'],\n",
    "    ['B cells', 'Dendritic Cells'],\n",
    "    ['B cells', 'Dendritic Cells']\n",
    "               ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.7370747873066266\n",
      "0.7494076802664078\n",
      "0.5763699102446633\n"
     ]
    }
   ],
   "source": [
    "\n",
    "############################################################################################\n",
    "# pbmc only\n",
    "############################################################################################\n",
    "\n",
    "\n",
    "vae = VAE(pbmc.nb_genes, n_batch=pbmc.n_batches, n_labels=pbmc.n_labels,\n",
    "          n_hidden=128, n_latent=10, n_layers=1, dispersion='gene')\n",
    "\n",
    "import torch\n",
    "trainer = UnsupervisedTrainer(vae, pbmc, train_size=1.0)\n",
    "# trainer.train(n_epochs=200)\n",
    "# torch.save(trainer.model,'../DE/pbmc.vae.model.pkl')\n",
    "trainer.model = torch.load('../DE/pbmc.vae.model.pkl')\n",
    "full = trainer.create_posterior(trainer.model, pbmc, indices=np.arange(len(pbmc)))\n",
    "latent, batch_indices, labels = full.sequential().get_latent()\n",
    "keys = pbmc.cell_types\n",
    "\n",
    "cell_type_label = [[np.where(pbmc.cell_types == x[i])[0].astype('int')[0] for i in [0, 1]] for x in comparisons]\n",
    "for t, comparison in enumerate(comparisons):\n",
    "    # Now for each comparison, let us create a posterior object and compute a Bayes factor\n",
    "    gene_set = gene_sets[t]\n",
    "    cell_indices = np.where(np.logical_or(\n",
    "        labels == cell_type_label[t][0],\n",
    "        labels == cell_type_label[t][1]))[0]\n",
    "    de_posterior = trainer.create_posterior(trainer.model, pbmc, indices=cell_indices)\n",
    "    scale_pbmc = de_posterior.sequential().get_harmonized_scale(0)\n",
    "    bayes_pbmc = get_bayes_factors(scale_pbmc,\n",
    "                                   pbmc.labels.ravel()[cell_indices],\n",
    "                                   cell_type_label[t][0],\n",
    "                                   cell_type_label[t][1])\n",
    "    print(auc_score_threshold(gene_set, bayes_pbmc, pbmc.gene_names))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "############################################################################################\n",
    "# all_dataset\n",
    "############################################################################################\n",
    "# VAE\n",
    "############################################################################################\n",
    "vae = VAE(all_dataset.nb_genes, n_batch=all_dataset.n_batches, n_labels=all_dataset.n_labels,\n",
    "          n_hidden=128, n_latent=10, n_layers=2, dispersion='gene')\n",
    "\n",
    "import torch\n",
    "trainer = UnsupervisedTrainer(vae, all_dataset, train_size=1.0)\n",
    "# trainer.train(n_epochs=200)\n",
    "# torch.save(trainer.model,'../DE/vae.model.rmdeT.pkl')\n",
    "trainer.model = torch.load('../DE/vae.model.rmdeT.pkl')\n",
    "\n",
    "trainer.train_set.entropy_batch_mixing()\n",
    "full = trainer.create_posterior(trainer.model, all_dataset, indices=np.arange(len(all_dataset)))\n",
    "latent, batch_indices, labels = full.sequential().get_latent()\n",
    "keys = all_dataset.cell_types\n",
    "\n",
    "\n",
    "from scvi.inference.posterior import entropy_batch_mixing\n",
    "sample = select_indices_evenly(2000, batch_indices)\n",
    "batch_entropy = entropy_batch_mixing(latent[sample, :], batch_indices[sample])\n",
    "\n",
    "latent_labelled = latent[batch_indices.ravel()==0, :]\n",
    "latent_unlabelled = latent[batch_indices.ravel()==1, :]\n",
    "labels_labelled = labels[batch_indices.ravel()==0]\n",
    "labels_unlabelled = labels[batch_indices.ravel()==1]\n",
    "n_labels = np.sum(batch_indices.ravel()==1)\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "neigh = KNeighborsClassifier(n_neighbors=10)\n",
    "neigh = neigh.fit(latent_labelled, labels_labelled)\n",
    "\n",
    "vae_pred = neigh.predict(latent)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['B cells' 'CD8 T cells' 'CD4 T cells' 'CD34 cells' 'NK cells'\n",
      " 'Dendritic Cells' 'CD14+ Monocytes']\n",
      "['CD8 T cells' 'CD4 T cells' 'CD34 cells' 'FCGR3A+ Monocytes' 'NK cells'\n",
      " 'Dendritic Cells' 'CD14+ Monocytes']\n"
     ]
    }
   ],
   "source": [
    "############################################################################################\n",
    "#  SCANVI\n",
    "############################################################################################\n",
    "scanvi = SCANVI(all_dataset.nb_genes, all_dataset.n_batches, all_dataset.n_labels, n_layers=2)\n",
    "scanvi.load_state_dict(trainer.model.state_dict(), strict=False)\n",
    "trainer_scanvi = SemiSupervisedTrainer(scanvi, all_dataset, classification_ratio=50,\n",
    "                                       n_epochs_classifier=1, lr_classification=5 * 1e-3)\n",
    "\n",
    "trainer_scanvi.labelled_set = trainer_scanvi.create_posterior(\n",
    "    indices=(all_dataset.batch_indices.ravel() == 0)\n",
    ")\n",
    "trainer_scanvi.unlabelled_set = trainer_scanvi.create_posterior(\n",
    "    indices=(all_dataset.batch_indices.ravel() == 1)\n",
    ")\n",
    "\n",
    "# trainer_scanvi.train(n_epochs=50)\n",
    "# torch.save(trainer_scanvi.model,'../DE/scanvi.model.pkl')\n",
    "trainer_scanvi.model = torch.load('../DE/scanvi.model.pkl')\n",
    "full = trainer_scanvi.create_posterior(trainer_scanvi.model, all_dataset, indices=np.arange(len(all_dataset)))\n",
    "latent, batch_indices, labels = full.sequential().get_latent()\n",
    "keys = all_dataset.cell_types\n",
    "pred = full.sequential().compute_predictions()\n",
    "\n",
    "# #todo map the other labels to the same labels as pbmc8k labels\n",
    "i = np.arange(len(keys))[keys=='CD8 T cells Naive']\n",
    "j = np.arange(len(keys))[keys=='CD8 T cells']\n",
    "pred[0][pred[0]==i]=j\n",
    "\n",
    "for CD4sub in ['CD4 T cells Naive','CD4 T cells Regulatory', 'CD4 T cells Memory', 'CD4 T cells 2']:\n",
    "    i = np.arange(len(keys))[keys == CD4sub]\n",
    "    j = np.arange(len(keys))[keys == 'CD4 T cells']\n",
    "    pred[0][pred[0] == i] = j\n",
    "    pred[1][pred[1]==i]=j\n",
    "\n",
    "print(keys[np.unique(pred[0][batch_indices.ravel()==1])])\n",
    "print(keys[np.unique(pred[1][batch_indices.ravel()==1])])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "from sklearn.metrics import confusion_matrix\n",
    "shared = set(keys[np.unique(pred[0][batch_indices.ravel()==1])]).intersection(set(keys[np.unique(pred[0][batch_indices.ravel()==0])]))\n",
    "shared = list(shared)\n",
    "CM_VAE = confusion_matrix(keys[pred[0][batch_indices.ravel()==1]],keys[vae_pred[batch_indices.ravel()==1]],labels=shared)\n",
    "CM_VAE = pd.DataFrame(CM_VAE,index=shared,columns=shared)\n",
    "WeightedAccuracy(pred[0][batch_indices.ravel()==0],vae_pred[batch_indices.ravel()==0],all_dataset.cell_types)\n",
    "WeightedAccuracy(pred[0][batch_indices.ravel()==1],vae_pred[batch_indices.ravel()==1],all_dataset.cell_types)\n",
    "\n",
    "CM_SCANVI = confusion_matrix(keys[pred[0][batch_indices.ravel()==1]],keys[pred[1][batch_indices.ravel()==1]],labels=shared)\n",
    "CM_SCANVI = pd.DataFrame(CM_SCANVI,index=shared,columns=shared)\n",
    "WeightedAccuracy(pred[0][batch_indices.ravel()==0],pred[1][batch_indices.ravel()==0],all_dataset.cell_types)\n",
    "WeightedAccuracy(pred[0][batch_indices.ravel()==1],pred[1][batch_indices.ravel()==1],all_dataset.cell_types)\n",
    "\n",
    "\n",
    "sample = select_indices_evenly(2000, batch_indices)\n",
    "batch_entropy = entropy_batch_mixing(latent[sample, :], batch_indices[sample])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>B cells</th>\n",
       "      <th>CD8 T cells</th>\n",
       "      <th>CD4 T cells</th>\n",
       "      <th>NK cells</th>\n",
       "      <th>Dendritic Cells</th>\n",
       "      <th>CD14+ Monocytes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>B cells</th>\n",
       "      <td>3663</td>\n",
       "      <td>3</td>\n",
       "      <td>1958</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8 T cells</th>\n",
       "      <td>57</td>\n",
       "      <td>8203</td>\n",
       "      <td>28640</td>\n",
       "      <td>82</td>\n",
       "      <td>9</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD4 T cells</th>\n",
       "      <td>15</td>\n",
       "      <td>262</td>\n",
       "      <td>10797</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK cells</th>\n",
       "      <td>8</td>\n",
       "      <td>4085</td>\n",
       "      <td>299</td>\n",
       "      <td>4050</td>\n",
       "      <td>3</td>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dendritic Cells</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>0</td>\n",
       "      <td>542</td>\n",
       "      <td>671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD14+ Monocytes</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1944</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 B cells  CD8 T cells  CD4 T cells  NK cells  Dendritic Cells  \\\n",
       "B cells             3663            3         1958         2                8   \n",
       "CD8 T cells           57         8203        28640        82                9   \n",
       "CD4 T cells           15          262        10797         1               14   \n",
       "NK cells               8         4085          299      4050                3   \n",
       "Dendritic Cells        1            1           61         0              542   \n",
       "CD14+ Monocytes        0            0            0         0                1   \n",
       "\n",
       "                 CD14+ Monocytes  \n",
       "B cells                        4  \n",
       "CD8 T cells                   53  \n",
       "CD4 T cells                   20  \n",
       "NK cells                     146  \n",
       "Dendritic Cells              671  \n",
       "CD14+ Monocytes             1944  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CM_VAE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>B cells</th>\n",
       "      <th>CD8 T cells</th>\n",
       "      <th>CD4 T cells</th>\n",
       "      <th>NK cells</th>\n",
       "      <th>Dendritic Cells</th>\n",
       "      <th>CD14+ Monocytes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>B cells</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3833</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>1738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8 T cells</th>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>11780</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>21428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD4 T cells</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>640</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK cells</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8760</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dendritic Cells</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>65</td>\n",
       "      <td>16</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD14+ Monocytes</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 B cells  CD8 T cells  CD4 T cells  NK cells  Dendritic Cells  \\\n",
       "B cells                0            0         3833         0                7   \n",
       "CD8 T cells            0            5        11780         2                0   \n",
       "CD4 T cells            0            3          640         1                1   \n",
       "NK cells               0            0         8760         0                0   \n",
       "Dendritic Cells        0            0           21        65               16   \n",
       "CD14+ Monocytes        0            0            5        22                0   \n",
       "\n",
       "                 CD14+ Monocytes  \n",
       "B cells                     1738  \n",
       "CD8 T cells                21428  \n",
       "CD4 T cells                 5200  \n",
       "NK cells                       2  \n",
       "Dendritic Cells               50  \n",
       "CD14+ Monocytes                1  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CM_SCANVI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "############################################################################################\n",
    "#  Computing AUC\n",
    "############################################################################################\n",
    "cell_type_label = [[np.where(all_dataset.cell_types == x[i])[0].astype('int')[0] for i in [0, 1]] for x in comparisons]\n",
    "for t, comparison in enumerate(comparisons):\n",
    "    # Now for each comparison, let us create a posterior object and compute a Bayes factor\n",
    "    gene_set = gene_sets[t]\n",
    "    cell_indices = np.where(np.logical_or(\n",
    "        pred == cell_type_label[t][0],\n",
    "        pred == cell_type_label[t][1]))[0]\n",
    "    de_posterior = trainer.create_posterior(trainer.model, all_dataset, indices=cell_indices)\n",
    "    scale_pbmc = de_posterior.sequential().get_harmonized_scale(0)\n",
    "    scale_68k = de_posterior.sequential().get_harmonized_scale(1)\n",
    "    # For Chenling: I looked again at the number of cells,\n",
    "    # if we use all of them, we are OK using just one sample from the posterior\n",
    "    # first grab the original bayes factor by ignoring the unlabeled cells\n",
    "    bayes_pbmc = get_bayes_factors(scale_pbmc,\n",
    "                                   all_dataset.labels.ravel()[cell_indices],\n",
    "                                   cell_type_label[t][0],\n",
    "                                   cell_type_label[t][1])\n",
    "    # second get them for all the predicted labels cross-datasets\n",
    "    probs_all_imputed_pbmc = get_bayes_factors(scale_pbmc,\n",
    "                                               pred[cell_indices],\n",
    "                                               cell_type_label[t][0],\n",
    "                                               cell_type_label[t][1], logit=False)\n",
    "    probs_all_imputed_68k = get_bayes_factors(scale_68k,\n",
    "                                              pred[cell_indices],\n",
    "                                              cell_type_label[t][0],\n",
    "                                              cell_type_label[t][1], logit=False)\n",
    "    p_s = pbmc.labels.shape[0] / all_dataset.labels.shape[0]\n",
    "    bayes_all_imputed = p_s * probs_all_imputed_pbmc + (1 - p_s) * probs_all_imputed_68k\n",
    "    bayes_all_imputed = np.log(bayes_all_imputed + 1e-8) - np.log(1 - bayes_all_imputed + 1e-8)\n",
    "    # COMMENT FOR CHENLING: the bayes variables (bayes_pbmc and bayes_all_imputed) should now span all values\n",
    "    # (not only positive or negative).\n",
    "    # Can you check that with np.min and np.max ?\n",
    "    print(auc_score_threshold(gene_set, bayes_pbmc, all_gene_symbols))\n",
    "    print(auc_score_threshold(gene_set, bayes_all_imputed, all_gene_symbols))\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Environment (conda_pytorch_p36)",
   "language": "python",
   "name": "conda_pytorch_p36"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
